File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/CoverageFunctionProfitService.java
package cz.fb.manaus.reactor.profit.progress;

import com.google.common.collect.ImmutableList;
import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.model.ProfitRecord;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.reactor.profit.progress.function.ProgressFunction;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.OptionalDouble;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@Component
public class CoverageFunctionProfitService extends AbstractFunctionProfitService implements FunctionProfitRecordCalculator {

    public CoverageFunctionProfitService(List<ProgressFunction> functions) {
        super(functions);
    }

    public List<ProfitRecord> getProfitRecords(List<SettledBet> bets, Optional<String> funcName,
                                               double chargeRate, Optional<String> projection) {
        return getProfitRecords(this::getProfitRecords, bets, chargeRate, funcName, projection);
    }

    @Override
    public List<ProfitRecord> getProfitRecords(ProgressFunction function, List<SettledBet> bets,
                                               BetCoverage coverage, Map<String, Double> charges) {
        Predicate<SettledBet> coveredPredicate = bet -> coverage.isCovered(bet.getBetAction().getMarket().getId(),
                bet.getSelectionId());
        var coverSeparation = bets.stream()
                .collect(Collectors.partitioningBy(coveredPredicate));
        var covered = coverSeparation.get(true);
        var solo = coverSeparation.get(false);

        var growthSeparation = covered.stream()
                .collect(Collectors.partitioningBy(this::isChargeGrowth));
        var head = growthSeparation.get(true);
        var tail = growthSeparation.get(false);

        var builder = ImmutableList.<ProfitRecord>builder();

        addRecord("solo", solo, function, coverage, charges)
                .ifPresent(builder::add);
        addRecord("covered", covered, function, coverage, charges)
                .ifPresent(builder::add);
        addRecord("covHead", head, function, coverage, charges)
                .ifPresent(builder::add);
        addRecord("covTail", tail, function, coverage, charges)
                .ifPresent(builder::add);
        return builder.build();
    }

    private Optional<ProfitRecord> addRecord(String categoryName, List<SettledBet> bets, ProgressFunction function,
                                             BetCoverage coverage, Map<String, Double> charges) {
        var average = getAverage(bets, function);
        var category = getValueCategory(function.getName() + "_" + categoryName, average);
        if (average.isPresent()) {
            return Optional.of(computeFunctionRecord(category, bets.stream(), charges, coverage));
        }
        return Optional.empty();
    }

    private boolean isChargeGrowth(SettledBet bet) {
        var action = bet.getBetAction();
        var chargeGrowth = action.getDoubleProperty("chargeGrowth");
        return chargeGrowth.orElse(Double.MAX_VALUE) > 1;
    }

    private OptionalDouble getAverage(List<SettledBet> bets, ProgressFunction function) {
        return bets.stream()
                .map(function)
                .filter(OptionalDouble::isPresent)
                .mapToDouble(OptionalDouble::getAsDouble)
                .average();
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/CoverageFunctionProfitService.kt
package cz.fb.manaus.reactor.profit.progress

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.model.ProfitRecord
import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.reactor.profit.progress.function.ProgressFunction
import org.springframework.stereotype.Component

@Component
class CoverageFunctionProfitService(functions: List<ProgressFunction>) : AbstractFunctionProfitService(functions), FunctionProfitRecordCalculator {

    fun getProfitRecords(bets: List<RealizedBet>, funcName: String?,
                         chargeRate: Double, projection: String? = null): List<ProfitRecord> {
        return getProfitRecords(this, bets, chargeRate, funcName, projection)
    }

    override fun getProfitRecords(function: ProgressFunction,
                                  bets: List<RealizedBet>,
                                  coverage: BetCoverage,
                                  charges: Map<String, Double>): List<ProfitRecord> {
        val (covered, solo) = bets.partition { coverage.isCovered(it.market.id, it.settledBet.selectionId) }

        val (head, tail) = covered.partition { this.isChargeGrowth(it) }

        val result = mutableListOf<ProfitRecord>()

        addRecord("solo", solo, function, coverage, charges)?.let { result.add(it) }
        addRecord("covered", covered, function, coverage, charges)?.let { result.add(it) }
        addRecord("covHead", head, function, coverage, charges)?.let { result.add(it) }
        addRecord("covTail", tail, function, coverage, charges)?.let { result.add(it) }
        return result
    }

    private fun addRecord(categoryName: String, bets: List<RealizedBet>, function: ProgressFunction,
                          coverage: BetCoverage, charges: Map<String, Double>): ProfitRecord? {
        val average = getAverage(bets, function)
        val category = getValueCategory(function.name + "_" + categoryName, average)
        return when (average) {
            null -> null
            else -> computeFunctionRecord(category, bets, charges, coverage)
        }
    }

    private fun isChargeGrowth(bet: RealizedBet): Boolean {
        val action = bet.betAction
        val chargeGrowth = action.properties["chargeGrowth"]
        return if (chargeGrowth != null) chargeGrowth.toDouble() > 1 else true
    }

    private fun getAverage(bets: List<RealizedBet>, function: ProgressFunction): Double? {
        return if (bets.isEmpty())
            return null
        else
            bets.mapNotNull { function(it) }
                    .average()
    }

}
